libraries/AC_Avoidance/AP_Astar_RT.cpp

```diff
@@ -0,0 +1,686 @@
+    }
+
+    if (_singleton != nullptr) {
+        AP_HAL::panic("AP_OART must be singleton");
+    }
+    _singleton = this;
+
+}
+
+// run background task to find best path and update avoidance_results
+// returns true and updates origin_new and destination_new if a best path has been found
+AP_OART_AStar::AP_OART_Astar_State AP_OART_AStar::update(const Location& current_loc, const Location& destination, const Vector2f &ground_speed_vec, Location &origin_new, Location &destination_new, bool proximity_only)
+{
+    const uint32_t start_ms = AP_HAL::millis();
+
+    // bendy ruler always sets origin to current_loc
+    origin_new = current_loc;
+
+    Vector3f current_pos,goal_pos;
+    if (!current_loc.get_vector_from_origin_NEU(current_pos) || !destination.get_vector_from_origin_NEU(goal_pos)) {
+        return RT_ASTAR_STATE_ERROR;
+    }
+    _path_destination = goal_pos;
+    _path_source = current_pos;
+
+    // create nodes from OA DB points
+    if (!create_oa_database_nodes()) {
+        if (_error_last_id == AP_OA_RTAStar_Error::RT_ASTAR_ERROR_NONE) {
+            // no points in the OADB
+            return RT_ASTAR_STATE_NOT_REQUIRED;
+        }
+        Write_RTAstar(RT_ASTAR_STATE_NOT_REQUIRED, _error_last_id, destination_new, 0, 0, 0, (AP_HAL::millis()-start_ms));
+        return RT_ASTAR_STATE_ERROR;
+    }
+
+    // time taken to create nodes
+    const uint32_t nodes_time = AP_HAL::millis() - start_ms;
+
+    // create visibility graph of OADB nodes
+    if (!create_oa_db_visgraph()) {
+        _error_last_id = AP_OA_RTAStar_Error::RT_ASTAR_ERROR_OUT_OF_MEMORY;
+        Write_RTAstar(RT_ASTAR_STATE_NOT_REQUIRED, _error_last_id, destination_new, nodes_time, 0, 0, (AP_HAL::millis()-start_ms));
+        return RT_ASTAR_STATE_ERROR;
+    }
+
+    // create visibility graph of goal and destination
+    if (!create_goal_and_destination_visgraph(current_pos,goal_pos)) {
+        _error_last_id = AP_OA_RTAStar_Error::RT_ASTAR_ERROR_OUT_OF_MEMORY;
+        Write_RTAstar(RT_ASTAR_STATE_NOT_REQUIRED, _error_last_id, destination_new, nodes_time, 0, 0, (AP_HAL::millis()-start_ms));
+        return RT_ASTAR_STATE_ERROR;
+    }
+
+    // time taken to create visgraph
+    const uint32_t visgraph_time = AP_HAL::millis() - nodes_time;
+
+    // run A-Star on all the visibility graphs
+    if (!find_shortest_path()) {
+        Write_RTAstar(RT_ASTAR_STATE_NOT_REQUIRED, _error_last_id, destination_new, nodes_time, visgraph_time, 0, (AP_HAL::millis()-start_ms));
+        return RT_ASTAR_STATE_ERROR;
+    }
+
+    // time taken to find a path
+    const uint32_t path_finding_time = AP_HAL::millis() - visgraph_time;
+
+    if (_path_numpoints <= 2) {
+        // direct path exists
+        Write_RTAstar(RT_ASTAR_STATE_NOT_REQUIRED, _error_last_id, destination_new, nodes_time, visgraph_time, path_finding_time, (AP_HAL::millis()-start_ms));
+        return RT_ASTAR_STATE_NOT_REQUIRED;
+    }
+
+    Vector3f dest_pos;
```

Comment:
maybe put a comment above this block of code to clarify what's going on below.  probably, "// path has been created, return latest point" or something similar.

As a side note, I've never been happy with how the path is returned in Dijkstra's.  It was a bit of a hack to make it fit in better with BendyRuler which can only return a single destination.  A better solution would have been to provide an interface to the path and then letter the caller do what they want with it.  For example SCurves would use the next two points to calculate a smooth path including the corner.  Anyway, this is not a requirement but I wanted to take this opportunity to point out shortcomings in Dijkstra's that also appear here.